API/src/main/java/io/github/jwdeveloper/tiktok/data/events/TikTokLinkMicBattleEvent.java

```diff
@@ -22,29 +22,153 @@
  */
 package io.github.jwdeveloper.tiktok.data.events;
 
-import io.github.jwdeveloper.tiktok.annotations.EventMeta;
-import io.github.jwdeveloper.tiktok.annotations.EventType;
+import io.github.jwdeveloper.tiktok.annotations.*;
 import io.github.jwdeveloper.tiktok.data.events.common.TikTokHeaderEvent;
-import io.github.jwdeveloper.tiktok.data.models.LinkMicBattleTeam;
+import io.github.jwdeveloper.tiktok.data.models.Picture;
+import io.github.jwdeveloper.tiktok.data.models.users.User;
 import io.github.jwdeveloper.tiktok.messages.webcast.WebcastLinkMicBattle;
 import lombok.Getter;
 
-import java.util.List;
+import java.util.*;
 
 /**
- * Triggered every time a battle starts.
+ * Triggered every time a battle starts & ends
  */
 @Getter
 @EventMeta(eventType = EventType.Message)
 public class TikTokLinkMicBattleEvent extends TikTokHeaderEvent {
-  private final Long battleId;
-  private final List<LinkMicBattleTeam> team1;
-  private final List<LinkMicBattleTeam> team2;
-
-  public TikTokLinkMicBattleEvent(WebcastLinkMicBattle msg) {
-    super(msg.getCommon());
-    battleId = msg.getId();
-    team1 = msg.getTeams1List().stream().map(LinkMicBattleTeam::new).toList();
-    team2 = msg.getTeams2List().stream().map(LinkMicBattleTeam::new).toList();
-  }
-}
+    private final Long battleId;
+    private final Integer battleStatus;
+    private final List<Team> teams;
+
+    public TikTokLinkMicBattleEvent(WebcastLinkMicBattle msg) {
+        super(msg.getCommon());
+        battleId = msg.getId();
+        battleStatus = msg.getBattleStatus();
+        teams = new ArrayList<>();
+        if (msg.getHostTeamCount() == 2) { // 1v1 battle
+            teams.add(new Team1v1(msg.getHostTeam(0), msg));
+            teams.add(new Team1v1(msg.getHostTeam(1), msg));
+        } else { // 2v2 battle
+            if (isFinished()) {
+                teams.add(new Team2v2(msg.getHostData2V2List().stream().filter(data -> data.getTeamNumber() == 1).findFirst().orElse(null), msg));
+                teams.add(new Team2v2(msg.getHostData2V2List().stream().filter(data -> data.getTeamNumber() == 2).findFirst().orElse(null), msg));
+            } else {
+                teams.add(new Team2v2(msg.getHostTeam(0), msg.getHostTeam(1), msg));
+                teams.add(new Team2v2(msg.getHostTeam(2), msg.getHostTeam(3), msg));
+            }
+        }
+
+        // Info:
+        // - msg.getDetailsList() & msg.getViewerTeamList() both only have content when battle is finished
+        // - msg.getDetailsCount() & msg.getViewerTeamCount() always is 2 only when battle is finished
+        // - msg.getHostTeamCount() always is 2 for 1v1 or 4 for 2v2
+    }
+
+    /**
+     battleStatus of 4 is Ongoing battle & 5 is Finished Battle
+     @return true if battle is finished otherwise false
+     */
+    public boolean isFinished() {
+        return battleStatus == 5;
+    }
+
+    public abstract static class Team {
```

Comment:
- Please avoid neasted classes for 2 reasons.
1. not handy to use.  example new TikTok.SomeData() instead of new SomeData()
2. We want to have inner TikTokLiveJava data classes as versatile as possible,
But now it only limits usecase of Team class to  TikTokLinkMicBattleEvent event.
Maybe I'm wrong but I have feeling that this class can be mapped from more events

Taking everything in the consideration, please move inner classes to own
files if they meet one of those points

Reply:
Fixed!

Reply:
> * Please avoid neasted classes for 2 reasons.
> 
> 1. not handy to use.  example new SomeData() is better then new TikTok.SomeData()
> 2. We want to have inner TikTokLiveJava data classes as versatile as possible,
>    But now it only limits usecase of Team class to  TikTokLinkMicBattleEvent event.
>    Maybe I'm wrong but I have feeling that this class can be mapped from more events in future
> 
> Taking everything in the consideration, please move inner classes to own files if they meet one of those points

As to point no1, you can `import com.example.MyClass.InnerClass` and use `new InnerClass()` that way ;)

Though I definitely think extracting Team to it's own file is a good choice here, I just wanted to give you my two cents worth :)